cmake/configure/configure_superlu_mt.cmake

```
@@ -0,0 +1,22 @@
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2012 - 2020 by the deal.II authors
+##
+## This file is part of the deal.II library.
+##
+## The deal.II library is free software; you can use it, redistribute
+## it, and/or modify it under the terms of the GNU Lesser General
+## Public License as published by the Free Software Foundation; either
+## version 2.1 of the License, or (at your option) any later version.
+## The full text of the license can be found in the file LICENSE.md at
+## the top level directory of deal.II.
+##
+## ---------------------------------------------------------------------
+
+#
+# Configuration for the SuperLU-MT library:
+#
+
+SET(FEATURE_SUPERLU-MT_DEPENDS BLAS)
```

```suggestion
SET(FEATURE_SUPERLU-MT_DEPENDS LAPACK)
```
`BLAS`  is configured together with `LAPACK` and only `DEAL_II_WITH_LAPACK` exists as option.

---

I still get the same error as before even after setting the path to SuperLU-MT in the cmake prefix
```
-- Include /calculate/Paras_Work/PhD_Work/DEALII-DEVELOP/dealii/cmake/configure/configure_superlu_mt.cmake
CMake Warning at cmake/macros/macro_find_package.cmake:27 (_FIND_PACKAGE):
  By not providing "FindSUPERLU-MT.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "SUPERLU-MT", but CMake did not find one.

  Could not find a package configuration file provided by "SUPERLU-MT" with
  any of the following names:

    SUPERLU-MTConfig.cmake
    superlu-mt-config.cmake

  Add the installation prefix of "SUPERLU-MT" to CMAKE_PREFIX_PATH or set
  "SUPERLU-MT_DIR" to a directory containing one of the above files.  If
  "SUPERLU-MT" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  cmake/macros/macro_configure_feature.cmake:127 (FIND_PACKAGE)
  cmake/macros/macro_configure_feature.cmake:239 (FEATURE_FIND_EXTERNAL)
  cmake/configure/configure_superlu_mt.cmake:22 (CONFIGURE_FEATURE)
  cmake/macros/macro_verbose_include.cmake:19 (INCLUDE)
  CMakeLists.txt:124 (VERBOSE_INCLUDE)


-- DEAL_II_WITH_SUPERLU-MT has unmet external dependencies.
CMake Error at cmake/macros/macro_configure_feature.cmake:112 (MESSAGE):
  

  Could not find the superlu-mt library!

  Please ensure that a suitable superlu-mt library is installed on your
  computer.

  If the library is not at a default location, either provide some hints for
  autodetection,

      $ SUPERLU-MT_DIR="..." cmake <...>
      $ cmake -DSUPERLU-MT_DIR="..." <...>

  or set the relevant variables by hand in ccmake.
```
I fail to understand why the library is not found even after providing the absolute path.

---

Does `SUPERLU` provide a `CMake` `Config` file? Then, it should work as is. Otherwise, you need to write a `FindSUPERLU-MT.cmake` find similar to the ones in `cmake/modules`. `cmake/modules/FindASSIMP.cmake` might be a good point to start if you don't need to detect more fancy configuration options.

---

After adding the following `FindSUPERLU-MT.cmake` file,
```
SET(SUPERLU-MT_DIR "" CACHE PATH "An optional hint to a SuperLU-MT installation")
SET_IF_EMPTY(SUPERLU-MT_DIR "$ENV{SUPERLU-MT_DIR}")

DEAL_II_FIND_LIBRARY(SUPERLU-MT_LIB NAMES libsuperlu_mt_PTHREAD.a
  HINTS ${SUPERLU-MT_DIR}/lib
  PATH_PREFIXES lib${LIB_PREFIX} lib64 lib
  )

DEAL_II_FIND_PATH(SUPERLU-MT_INC slu_mt_ddefs.h
  HINTS ${SUPERLU-MT_DIR}
  PATH_SUFFIXES include
  )

DEAL_II_PACKAGE_HANDLE(SUPERLU-MT
  LIBRARIES REQUIRED SUPERLU-MT_LIB
  INCLUDE_DIRS REQUIRED SUPERLU-MT_INC
  USER_INCLUDE_DIRS REQUIRED SUPERLU-MT_INC
  CLEAR SUPERLU-MT_DIR_LIB SUPERLU-MT_INC
  )
```
We get this type of linker errors:
```
/usr/bin/ld.gold: error: /calculate/iwtm99_lokal/Paras_HomeOffice/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/superlu-mt-3.1-d5qq466ouorvdkeox332jcpnxb5fg4tt/lib/../lib/libsuperlu_mt_PTHREAD.a(get_perm_c.o): requires dynamic R_X86_64_PC32 reloc against 'getata' which may overflow at runtime; recompile with -fPIC
```
Any clues on how to fix this?

---

Maybe patch for the SuperLU-MT spack package is what you need:
[superlu-mt: add pic and int64 variants](https://github.com/spack/spack/pull/11361)

If it helps, then perhaps try to raise awareness there and push to get that PR merged.

---

Yes, that's essentially a bug in the SuperLU configuration. Or, more correctly, it is dangerous to link static (`.a`) libraries together with shared (`.so`) libraries. Does SuperLU not allow to compile itself into a shared library?

---

> Maybe patch for the SuperLU-MT spack package is what you need:
> [superlu-mt: add pic and int64 variants](https://github.com/spack/spack/pull/11361)
> 
> If it helps, then perhaps try to raise awareness there and push to get that PR merged.

Thank you so much for the hint. I did these changes on my local spack installation and it works now.